### PR TITLE
Allow client to configure parameters

### PR DESCRIPTION
**Pull request by sfc-gh-mnaides**

### CODE REVIEW

File: src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java [hunk 1]
```diff
@@ -314,7 +315,7 @@ void distributeFlushTasks() {
       }
 
       // Distribute work at table level, create a new blob if reaching the blob size limit
-      while (itr.hasNext() && totalBufferSize <= MAX_BLOB_SIZE_IN_BYTES) {
+      while (itr.hasNext() && totalBufferSize <= maxBlobSizeInBytes) {
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java:318]: Why this change if `maxBlobSizeInByes` is always initialized to the constant? Did you mean to use something from `owningClient` instead?

File: src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestChannelFactory.java [hunk 2]
```diff
@@ -26,6 +26,7 @@ static SnowflakeStreamingIngestChannelBuilder builder(String name) {
     private String encryptionKey;
     private Long encryptionKeyId;
     private OpenChannelRequest.OnErrorOption onErrorOption;
+    private ParameterProvider parameterProvider;
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestChannelFactory.java:29]: Is this initialized anywhere?

File: src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestClientInternal.java [hunk 3]
```diff
@@ -73,6 +73,9 @@
   // Counter to generate unique request ids per client
   private final AtomicLong counter = new AtomicLong(0);
 
+  // Provides constant values set by Snowflake server
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestClientInternal.java:76]: These aren't set by the server, right?

File: src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestClientInternal.java [hunk 4]
```diff
@@ -192,7 +201,7 @@ public SnowflakeStreamingIngestClientInternal(
    * @param name the name of the client
    */
   SnowflakeStreamingIngestClientInternal(String name) {
-    this(name, null, null, null, true, null);
+    this(name, null, null, null, true, null, null);
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestClientInternal.java:204]: Maybe just use an empty map here instead?

File: src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java (hunk 1 above)
Comment (sfc-gh-mnaides) [reply src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java:318]: I repurposed my branch that took parameters from the server (including this one) and made some mistakes in the cleanup, including this

File: src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestClientInternal.java (hunk 3 above)
Comment (sfc-gh-mnaides) [reply src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestClientInternal.java:76]: Right, fixing comment

File: src/main/java/net/snowflake/ingest/streaming/internal/ParameterProvider.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+      this.parameterMap.put(
+          key, parameterOverrides.getOrDefault(key, props.getOrDefault(key, defaultValue)));
+    } else if (parameterOverrides != null) {
+      this.parameterMap.put(key, parameterOverrides.getOrDefault(key, defaultValue));
+    } else if (props != null) {
+      this.parameterMap.put(key, props.getOrDefault(key, defaultValue));
+    }
+  }
+  /**
+   * Sets parameter values by first checking 1. parameterOverrides 2. props 3. default value
```
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/streaming/internal/ParameterProvider.java:45]: can you make it clear that props is the value from the profile.json file?

File: src/main/java/net/snowflake/ingest/streaming/internal/ParameterProvider.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+  /**
+   * Construct empty ParameterProvider that will supply default values until updated by
+   * client/configure API call
+   */
+  public ParameterProvider() {}
+
+  private void updateValue(
+      String key, Object defaultValue, Map<String, Object> parameterOverrides, Properties props) {
+    if (parameterOverrides != null && props != null) {
```
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/streaming/internal/ParameterProvider.java:35]: props should be always be not null, right? So we should be checking
1. parameterOverrides is null or not
2. parameterOverrides.get(key) is null or not
3. props.get(key) is null or not

File: src/main/java/net/snowflake/ingest/streaming/internal/ParameterProvider.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+  static final long INSERT_THROTTLE_THRESHOLD_IN_PERCENTAGE_DEFAULT = 5;
+
+  /** Map of parameter name to parameter value. This will be set by client/configure API Call. */
+  private Map<String, Object> parameterMap = new HashMap<>();
+
+  /**
+   * Construct empty ParameterProvider that will supply default values until updated by
+   * client/configure API call
+   */
+  public ParameterProvider() {}
```
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/streaming/internal/ParameterProvider.java:31]: we should just pass `parameterOverrides` and `props` to the construct, so both the `updateValue` and `setParameterMap` function could be simplified

File: src/main/java/net/snowflake/ingest/streaming/internal/StreamingIngestStage.java [hunk 8]
```diff
@@ -90,6 +94,11 @@
     this.role = role;
     this.requestBuilder = requestBuilder;
     this.clientName = clientName;
+    /*
+    All integer numbers will be deserialized as longs.
+    If this is false Jackson will deserialize to int or long based on size
+    */
+    this.mapper.configure(DeserializationFeature.USE_LONG_FOR_INTS, true);
```
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/streaming/internal/StreamingIngestStage.java:101]: everything in this file could be reverted.

File: src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestClientInternal.java [hunk 9]
```diff
@@ -167,6 +171,8 @@
       SharedMetricRegistries.add("Metrics", metrics);
     }
 
+    this.parameterProvider.setParameterMap(parameterOverrides, prop);
```
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestClientInternal.java:174]: Could we add a log after this so we could know the actual values being used in the end?

File: src/main/java/net/snowflake/ingest/utils/Utils.java [hunk 10]
```diff
@@ -74,6 +74,7 @@ public static void assertNotNull(String name, Object value) throws SFException {
    */
   public static Properties createProperties(Properties inputProp, boolean sslEnabled) {
     Properties properties = new Properties();
+    properties.putAll(inputProp);
```
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/utils/Utils.java:77]: I think we would still want to use the switch case below to add to the properties, it does more things like toLowerCase and log a entry if the key is not useful.

File: src/main/java/net/snowflake/ingest/streaming/internal/ParameterProvider.java (hunk 5 above)
Comment (sfc-gh-mnaides) [reply src/main/java/net/snowflake/ingest/streaming/internal/ParameterProvider.java:45]: The profile.json only really exists in our example, since the client factory just takes a Properties object

File: src/main/java/net/snowflake/ingest/utils/Utils.java (hunk 10 above)
Comment (sfc-gh-mnaides) [reply src/main/java/net/snowflake/ingest/utils/Utils.java:77]: Please double check my logic, but i think I kept all of the reformatting we do below and figured it was simpler to just pass along all the props even if we don't use them

Comment (sfc-gh-tzhang) [reply src/main/java/net/snowflake/ingest/utils/Utils.java:77]: hmmm, IIUC, you remove a few cases in the switch statement below, so let's say if the input prop is `USER` instead of `user`, we won't able to get it later anymore with our change, right? Also now we will log an entry saying user property is ignored which is not correct. I feel like the old logic is more clear, could we change it back?
